File: examples/csharp/flappy_goud/GameManager.cs
// GameManager.cs

using System;
using System.Collections.Generic;
using GoudEngine.Input;

public class GameManager
{
    private GoudGame game;

    private ScoreCounter scoreCounter;
    private Bird bird;
    private List<Pipe> pipes;
    private float pipeSpawnTimer;

    // Texture IDs for immediate-mode rendering (64-bit handles)
    private ulong backgroundTextureId;
    private ulong baseTextureId;
    private ulong pipeTextureId;

    // Background dimensions (actual image: 288 × 512)
    private const float BackgroundWidth = 288f;
    private const float BackgroundHeight = 512f;

    // Base dimensions (actual image: 336 × 112)
    private const float BaseWidth = 336f;
    private const float BaseHeight = 112f;

    public GameManager(GoudGame game)
    {
        this.game = game;
        bird = new Bird(game);
        pipes = new List<Pipe>();
        pipeSpawnTimer = 0;
        scoreCounter = new ScoreCounter();
    }

    public void Initialize()
    {
        // Load all textures once at initialization (64-bit handles)
        backgroundTextureId = game.LoadTexture("assets/sprites/background-day.png");
        baseTextureId = game.LoadTexture("assets/sprites/base.png");
        pipeTextureId = game.LoadTexture("assets/sprites/pipe-green.png");

        bird.Initialize();
        scoreCounter.Initialize(game);
    }

    public void Start()
    {
        bird.Reset();
        pipes.Clear();
        scoreCounter.ResetScore();
        pipeSpawnTimer = 0;
    }

    public void Update(float deltaTime)
    {
        // If Escape is pressed, close the game
        if (game.IsKeyPressed(Keys.Escape))
        {
            game.Close();
            return;
        }

        // If R is pressed, restart the game
        if (game.IsKeyPressed(Keys.R))
        {
            ResetGame();
            return;
        }

        // Update Bird Movement with deltaTime
        bird.Update(deltaTime);

        // Check if bird is colliding with the base (ground)
        if (bird.Y + Bird.Height > GameConstants.ScreenHeight)
        {
            ResetGame();
            return;
        }

        // Check if bird is off-screen (top)
        if (bird.Y < 0)
        {
            ResetGame();
            return;
        }

        // Update Pipes with deltaTime and check for collisions
        foreach (var pipe in pipes)
        {
            pipe.Update(deltaTime);

            // Use AABB collision detection
            if (pipe.CollidesWithBird(bird.X, bird.Y, Bird.Width, Bird.Height))
            {
                ResetGame();
                return;
            }
        }

        // Spawn new pipes
        pipeSpawnTimer += deltaTime;
        if (pipeSpawnTimer > GameConstants.PipeSpawnInterval)
        {
            pipeSpawnTimer = 0;
            pipes.Add(new Pipe(game));
        }

        // Remove off-screen pipes and increment score
        pipes.RemoveAll(pipe =>
        {
            if (pipe.IsOffScreen())
            {
                scoreCounter.IncrementScore();
                return true;
            }
            return false;
        });

        // Draw everything (immediate-mode rendering)
        Draw();
    }

    /// <summary>
    /// Draws all game objects using immediate-mode rendering.
    /// Called each frame after Update.
    /// Note: DrawSprite uses center-based positioning, so we offset by half width/height.
    /// Draw order determines layering (later = on top).
    /// </summary>
    private void Draw()
    {
        // Layer 0: Background
        game.DrawSprite(backgroundTextureId, 
            BackgroundWidth / 2, 
            BackgroundHeight / 2, 
            BackgroundWidth, BackgroundHeight);

        // Layer 1: Score (behind pipes, in front of background)
        scoreCounter.Draw(game);

        // Layer 2: Pipes
        foreach (var pipe in pipes)
        {
            pipe.Draw(pipeTextureId);
        }

        // Layer 3: Bird (in front of pipes)
        bird.Draw();

        // Layer 4: Base/ground (in front of everything)
        game.DrawSprite(baseTextureId, 
            BaseWidth / 2, 
            GameConstants.ScreenHeight + BaseHeight / 2, 
            BaseWidth, BaseHeight);
    }

    private void ResetGame()
    {
        scoreCounter.ResetScore();
        Start(); // Restart the game
    }
}

File: examples/swift/flappy_bird/Sources/FlappyBird/GameManager.swift
// GameManager.swift

import Foundation
import GoudEngine

final class GameManager {
    private let game: GoudGame

    private let scoreCounter = ScoreCounter()
    private let bird: Bird
    private var pipes: [Pipe] = []
    private var pipeSpawnTimer: Float = 0

    private var backgroundTextureId: UInt64 = 0
    private var baseTextureId: UInt64 = 0
    private var pipeTextureId: UInt64 = 0

    private let backgroundWidth: Float = 288
    private let backgroundHeight: Float = 512
    private let baseWidth: Float = 336
    private let baseHeight: Float = 112

    init(game: GoudGame) {
        self.game = game
        bird = Bird(game: game)
    }

    func initialize() {
        backgroundTextureId = game.loadTexture(path: "assets/sprites/background-day.png")
        baseTextureId = game.loadTexture(path: "assets/sprites/base.png")
        pipeTextureId = game.loadTexture(path: "assets/sprites/pipe-green.png")

        bird.initialize()
        scoreCounter.initialize(game: game)
    }

    func start() {
        bird.reset()
        pipes.removeAll()
        scoreCounter.resetScore()
        pipeSpawnTimer = 0
    }

    func update(deltaTime: Float) {
        if game.isKeyPressed(key: .ESCAPE) {
            game.requestClose()
            return
        }

        if game.isKeyPressed(key: .R) {
            resetGame()
            return
        }

        let didFlap = bird.update(deltaTime: deltaTime)
        if didFlap {
            let _ = game.audioPlay(data: EmbeddedAudioClips.flapWav)
        }

        // Ground collision
        if bird.y + Bird.height > Float(GameConstants.screenHeight) {
            resetGame()
            return
        }

        // Top of screen
        if bird.y < 0 {
            resetGame()
            return
        }

        // Update pipes and check collisions
        for pipe in pipes {
            pipe.update(deltaTime: deltaTime)

            if pipe.collidesWithBird(
                birdX: bird.x, birdY: bird.y,
                birdWidth: Bird.width, birdHeight: Bird.height
            ) {
                resetGame()
                return
            }
        }

        // Spawn new pipes
        pipeSpawnTimer += deltaTime
        if pipeSpawnTimer > GameConstants.pipeSpawnInterval {
            pipeSpawnTimer = 0
            pipes.append(Pipe(game: game))
        }

        // Remove off-screen pipes and increment score
        pipes.removeAll { pipe in
            if pipe.isOffScreen() {
                scoreCounter.incrementScore()
                return true
            }
            return false
        }

        draw()
    }

    private func draw() {
        // Layer 0: Background
        game.drawSprite(
            texture: backgroundTextureId,
            x: backgroundWidth / 2,
            y: backgroundHeight / 2,
            width: backgroundWidth,
            height: backgroundHeight
        )

        // Layer 1: Score
        scoreCounter.draw(game: game)

        // Layer 2: Pipes
        for pipe in pipes {
            pipe.draw(pipeTextureId: pipeTextureId)
        }

        // Layer 3: Bird
        bird.draw()

        // Layer 4: Base/ground
        game.drawSprite(
            texture: baseTextureId,
            x: baseWidth / 2,
            y: Float(GameConstants.screenHeight) + baseHeight / 2,
            width: baseWidth,
            height: baseHeight
        )
    }

    private func resetGame() {
        let _ = game.audioPlay(data: EmbeddedAudioClips.resetWav)
        scoreCounter.resetScore()
        start()
    }
}
